### PR TITLE
Python3 compatible

### PR DESCRIPTION
**Pull request by zxkjack123**
This PR ports PyNE code, which currently does not support python3 `pymoab` or `dagmc-pymoab`, to python2/3 compatible with `pymoab` and `dagmc-pymoab`.
This work started by @kkiesling and @bam241 at this [branch](https://github.com/kkiesling/pyne/commits/python3).  Due to the reason that rebasing of her branch is even more difficult than just redo the commits on the current develop, I started this draft PR to continue this work. Some of the changes in that branch were used in this PR.

By following the tutorial for [Porting Python 2 Code to Python 3](https://docs.python.org/3/howto/pyporting.html), [Cheat Sheet: Writing Python 2-3 compatible code](https://python-future.org/compatible_idioms.html), changes are made according to the need ot passing the unit tests of the following test functions:
- [x] : test_alara.py
- [x] : test_mesh.py
- [x] : test_dagmc.py
- [x] : test_mcnp.py
- [x] : test_partisn.py
- [x] : test_r2s.py 
- [x] : test_source_sampling.py
- [x] : test_variancereduction.py
- [x] : test_utils.py

### CODE REVIEW

**Comment by zxkjack123** [comment]
Update: One of the test functions in `tests/test_variancereduction.py` is commented for the purpose of avoiding segmentation fault.

File: pyne/mcnp.py [hunk 1]
```diff
@@ -2281,7 +2281,7 @@ def _mesh_to_mat_cards(mesh, divs, frac_type):
     mat_cards = ""
     idx = mesh.iter_structured_idx('xyz')
     for i in idx:
-        mesh.mats[i].metadata['mat_number'] = i + 1
+        mesh.mats[i].metadata['mat_number'] = int(i + 1)
```
Comment (pshriwise) [inline pyne/mcnp.py:2284]: I'm a little confused about this one. Why wouldn't this be an int in either Python 2 or 3? If `idx` has non-`int`'s in it, that might be a bug in `mesh.iter_structured_idx`. (Now worried I'm going to be pointing out my own error here... :grimacing: )

File: pyne/mesh.py [hunk 2]
```diff
@@ -726,7 +727,7 @@ def __init__(self, mesh=None, structured=False,
                 # check for the structured box tag on the instance
                 try:
                     box_tag = self.mesh.tag_get_handle(_BOX_DIMS_TAG_NAME)
-                except types.MB_TAG_NOT_FOUND as e:
+                except Exception as e:
```
Comment (pshriwise) [inline pyne/mesh.py]: It's really better to capture specific exceptions to make sure we aren't overlooking an error that should be handled differently.

File: pyne/mesh.py [hunk 3]
```diff
@@ -1076,12 +1083,12 @@ def common_ve_tags(self, other):
                                   self.structured_set,
                                   types.MBMAXTYPE,
                                   dim=3)
-        self_tags = self.mesh.tag_get_tags_on_entity(self_it.next())
+        self_tags = self.mesh.tag_get_tags_on_entity(self_it.__next__())
```
Comment (pshriwise) [inline pyne/mesh.py]: Can we use `next(self_it)` like we did above for this? It would be considerably cleaner.

File: pyne/mesh.py [hunk 4]
```diff
@@ -1300,6 +1307,9 @@ def structured_get_divisions(self, dim):
         """
         self._structured_check()
 
+        ## sometimes the dim is the ascii of the 'x', 'y', 'z'
+        #if isinstance(dim, int):
+        #    dim = chr(dim)
```
Comment (pshriwise) [inline pyne/mesh.py]: If this is no longer the case, let's remove all these commented lines.

File: pyne/mesh.py [hunk 5]
```diff
@@ -1518,7 +1528,7 @@ def _structured_step_iter(it, n):
     Return the nth item in the iterator.
     """
     it.step(n)
-    r = it.next()
+    r = it.__next__()
```
Comment (pshriwise) [inline pyne/mesh.py]: Same comment as above about using `next()` over `.__next__()`

File: pyne/partisn.py [hunk 6]
```diff
@@ -529,7 +529,7 @@ def _write_input(title, block01, block02, block03, block04, block05, cards, file
 
     block1_cards = []
     if 'block1' in cards:
-        for card, value in cards['block1'].iteritems():
+        for card, value in sorted(cards['block1'].items()):
```
Comment (pshriwise) [inline pyne/partisn.py:532]: Why sorted here? Should we be using a different data structure?

File: pyne/variancereduction.py [hunk 7]
```diff
@@ -3,9 +3,13 @@
...
 import numpy as np
 from pyne.particle import mcnp
-from mcnp import Wwinp
+from .mcnp import Wwinp
 from pyne.mesh import Mesh, MeshError, HAVE_PYMOAB
-from itertools import izip
+# default zip in python3 behaves as itertools.izip as python2
+try:
+    from itertools import izip as zip
+except:
```
Comment (pshriwise) [inline pyne/variancereduction.py]: Let's except an `ImportError` specifically here

File: tests/files_test_partisn/partisn_1D_expected.inp [hunk 8]
```diff
@@ -30,7 +30,7 @@ t
 / 
 / Note: NAMES is not all inclusive. Only NAMES that are present in
 / meshed area are listed.
-names= He4 Hg202 Hg200 Hg201 Hg196 Hg204 He3 Hg198 Hg199 
+names= He3 He4 Hg196 Hg198 Hg199 Hg200 Hg201 Hg202 Hg204
```
Comment (pshriwise) [inline tests/files_test_partisn/partisn_1D_expected.inp:33]: Why the changes to these test results?

File: tests/test_mesh.py [hunk 9]
```diff
@@ -4,6 +4,17 @@
 import shutil
 import warnings
 import itertools
+# default zip in python3 behaves as itertools.izip as python2
+try:
+    from itertools import izip as zip
+except:
```
Comment (pshriwise) [inline tests/test_mesh.py]: Same as above `ImportError` only

**Review by pshriwise — CHANGES_REQUESTED** [review]
Thank you for this effort @zxkjack123! Just some line comments for now. I'll take a look at test changes on the next pass.

File: pyne/mesh.py (hunk 4 above)
Comment (zxkjack123) [reply pyne/mesh.py]: My bad, forgot to clean up here. Will change it in next commit.

File: pyne/partisn.py (hunk 6 above)
Comment (zxkjack123) [reply pyne/partisn.py:532]: Sort the content here is to make sure the output blocks have the same sequence in python2 and python3. Otherwise, they have different sequences, then the output file compare will return a `False`.

File: pyne/variancereduction.py (hunk 7 above)
Comment (zxkjack123) [reply pyne/variancereduction.py]: There is actually no an `ImportError`.
For python3, there is a buildin `zip`, we don't need to import.
For python2, we need to use `izip` (which works the same as the `zip` in python3.
This `try` here, works for only python2. I will remove the `except` here and add more description.

File: tests/files_test_partisn/partisn_1D_expected.inp (hunk 8 above)
Comment (zxkjack123) [reply tests/files_test_partisn/partisn_1D_expected.inp:33]: Because the output sequence of python2 and python3 are different.
There is a set storing these nuclide names, we iterate the set and print out the nucldie names.
However, the default iterate sequence for this set seems to be different in python2 and python3.

You can see that their content is actually the same, but just with different sequence. Well, the sequence is not important.
But for the compatibility, we need to sort and then output. The new results are sorted output, the old results not sorted.

File: pyne/mesh.py (hunk 2 above)
Comment (zxkjack123) [reply pyne/mesh.py]: I agree to use specific exceptions. But the problem is `types.MB_TAG_NOT_FOUND` exception is not recognized by python3. Reason not known yet.

File: pyne/variancereduction.py (hunk 7 above)
Comment (pshriwise) [reply pyne/variancereduction.py]: But in Python3 this will raise an exception, right?

Comment (zxkjack123) [reply pyne/variancereduction.py]: Yes, for python3 there will be an exception for import `izip`. But when that happens, it means we do have `zip` we want, then we dont't need to do anything with this exception.

File: pyne/mcnp.py (hunk 1 above)
Comment (zxkjack123) [reply pyne/mcnp.py:2284]: This `i` here is a `np.int32` object, when it add 1, it becomes `np.int64` object.
This is the behavior of `np.int64` difference in python2 and python3.
Assuming `a` is a `np.int64` object.
In python2, `isinstance(a, int)` is `True`, however, in python3, `isinstance(a, int)` returns `False`. 
Therefore, this line: https://github.com/pyne/pyne/blob/cd90c6f49ec58d47af7a1e03e91470d6516b12d3/pyne/jsoncpp.pyx#L63, behaves differently. Causing the problem.

There could be two ways to fix it:
- convert the `i+1` to int
- modify the type check in `jsoncpp.pyx`

I used the easy one here. But I think it deserves a specific PR to do the second option.

File: pyne/mesh.py (hunk 2 above)
Comment (pshriwise) [reply pyne/mesh.py]: Yeah I'll look into that and see if we can do something better here...

File: pyne/variancereduction.py (hunk 7 above)
Comment (pshriwise) [reply pyne/variancereduction.py]: Agreed. I think that handling the exception with `pass` is fine, but that we should only do that for the specific exception `ImportError`. 


```Python
try:
    from itertools import izip as zip
except ImportError:
    pass
```

File: tests/files_test_partisn/partisn_1D_expected.inp (hunk 8 above)
Comment (pshriwise) [reply tests/files_test_partisn/partisn_1D_expected.inp:33]: Gotcha. That makes sense and make the output more understandable

File: pyne/partisn.py (hunk 6 above)
Comment (pshriwise) [reply pyne/partisn.py:532]: I'm wondering if using an [OrderedDict](https://docs.python.org/3/library/collections.html#collections.OrderedDict) might be useful here.

File: pyne/variancereduction.py (hunk 7 above)
Comment (zxkjack123) [reply pyne/variancereduction.py]: Got it. Will make this change in the next update.

File: pyne/mesh.py (hunk 3 above)
Comment (zxkjack123) [reply pyne/mesh.py]: Sure, will apply this suggestion in the next update.

Comment (zxkjack123) [reply pyne/mesh.py]: Sadly, using `next` to replace `__next__()` will cause error.

Comment (pshriwise) [reply pyne/mesh.py]: This is because the `MeshSetIterator` isn't quite setup like a proper Python generator class. I've submitted a PR into your `python3_compatible` branch to fix that issue. Once that is merged, we can replace any other calls to `it.__next__()` with `next(it)`

File: pyne/mesh.py (hunk 2 above)
Comment (pshriwise) [reply pyne/mesh.py]: I think this exception can be captured as a `RuntimeError`.

```python console
In [1]: from pymoab import core, types                                                                                                            

In [2]: mb = core.Core()                                                                                                                          

In [3]: mb.tag_get_handle(types.CATEGORY_TAG_NAME)                                                                                                

pymoab/core.pyx in pymoab.core.Core.tag_get_handle()

pymoab/types.pyx in pymoab.types.check_error()

RuntimeError: MOAB ErrorCode: MB_TAG_NOT_FOUND
```

Comment (zxkjack123) [reply pyne/mesh.py]: Applied in 0df5d71.

File: tests/test_mesh.py [hunk 10]
```diff
@@ -4,6 +4,21 @@
...
+# For python3, do nothing with zip.
+try:
+    from itertools import izip as zip
+except ImportError:
+    pass
+
+# izip_longest in python3 was renamed to zip_longest in python3
+try:
+    from itertools import izip_longest as zip_longest
+except:
```
Comment (pshriwise) [inline tests/test_mesh.py]: We'll want to make sure this is only excepting an `ImportError` as well.

**Review by pshriwise — COMMENTED** [review]
One more exception to update. Also, @zxkjack123 did you have a chance to look into whether or not using an `OrderedDict` would help avoid all the `sorted` calls?

**Comment by zxkjack123** [comment]
@pshriwise Sure, I will change the dictionary to `OrderedDict` for input settings. There are also some dictionaries are used for nuclide names, it's better to use` sorted` to print.

**Comment by zxkjack123** [comment]
@pshriwise I decide to give up using `OrderedDict`. It works well with python3, but cause failure with python2,  255c4da failed the CI tests.

**Comment by pshriwise** [comment]
Sounds good. From the look of the interface we'd need to make sure that all `block*` values of the `cards` `dict` are ordered, which is probably as much of a hassle as using `sorted` where we need it. I'm good w/ moving on since this file write is hopefully a one-off operation in and isn't too expensive in the context of a full workflow.

**Review by pshriwise — APPROVED** [review]
Looking forward to Python3 compatibility! Thanks @zxkjack123!

**Comment by pshriwise** [comment]
Going to merge on Monday if there are no other comments on this.
